edge/pkg/devicetwin/dtclient/device_db.go

```diff
@@ -89,58 +89,98 @@ func UpdateDeviceMulti(updates []DeviceUpdate) error {
 func AddDeviceTrans(adds []Device, addAttrs []DeviceAttr, addTwins []DeviceTwin) error {
 	var err error
 	obm := dbm.DBAccess
-	obm.Begin()
+	err = obm.Begin()
+	if err != nil {
+		klog.Errorf("begin transaction failed: %v", err)
+		return err
+	}
 	for _, add := range adds {
 		err = SaveDevice(&add)
 
 		if err != nil {
 			klog.Errorf("save device failed: %v", err)
-			obm.Rollback()
+			errRollback := obm.Rollback()
+			if errRollback != nil {
+				klog.Errorf("transaction rollback failed: %v", errRollback)
+				return errRollback
+			}
 			return err
 		}
 	}
 
 	for _, attr := range addAttrs {
 		err = SaveDeviceAttr(&attr)
 		if err != nil {
-			obm.Rollback()
+			errRollback := obm.Rollback()
+			if errRollback != nil {
+				klog.Errorf("transaction rollback failed: %v", errRollback)
+				return errRollback
```

Comment:
I think we should return the SaveDeviceAttr err,  it is the main err that why AddDeviceTrans exit, so  ignore the rollback err is better

Reply:
you are right，**no matter whether the rollback is successful or not**, the transaction will not be committed, so there is no need to handle the rollback error .
i will close this pr